xarray/ufuncs.py

```diff
@@ -0,0 +1,125 @@
+        for node in obj.subtree:
+            _walk_array_namespaces(node.dataset, namespaces)
+    elif isinstance(obj, xr.Dataset):
+        for name in obj.data_vars:
+            _walk_array_namespaces(obj[name], namespaces)
+    elif isinstance(obj, GroupBy):
+        _walk_array_namespaces(next(iter(obj))[1], namespaces)
+    elif isinstance(obj, xr.DataArray | xr.Variable):
+        _walk_array_namespaces(obj.data, namespaces)
+    elif isinstance(obj, array_type("dask")):
+        _walk_array_namespaces(obj._meta, namespaces)
+    else:
+        namespace = getattr(obj, "__array_namespace__", None)
+        if namespace is not None:
+            namespaces.add(namespace())
+
+    return namespaces
+
+
+class _UFuncDispatcher:
+    """Wrapper for dispatching ufuncs."""
+
+    def __init__(self, name):
+        self._name = name
+
+    def __call__(self, *args, **kwargs):
+        xps = set()
+        for arg in args:
+            _walk_array_namespaces(arg, xps)
+
+        xps.discard(np)
+        if len(xps) > 1:
+            names = [module.__name__ for module in xps]
+            raise ValueError(
+                f"Mixed array types {names} are not supported by xarray.ufuncs"
+            )
+
+        xp = next(iter(xps)) if len(xps) else np
+        func = getattr(xp, self._name, None)
+
+        if func is None:
+            warnings.warn(
+                f"Function {self._name} not found in {xp.__name__}, falling back to numpy",
+                stacklevel=2,
+            )
+            func = getattr(np, self._name)
+
+        return xr.apply_ufunc(func, *args, dask="parallelized", **kwargs)
+
+
+def _skip_signature(doc, name):
+    if not isinstance(doc, str):
+        return doc
+
+    if doc.startswith(name):
+        signature_end = doc.find("\n\n")
+        doc = doc[signature_end + 2 :]
+
+    return doc
+
+
+def _remove_unused_reference_labels(doc):
+    if not isinstance(doc, str):
+        return doc
+
+    max_references = 5
+    for num in range(max_references):
+        label = f".. [{num}]"
+        reference = f"[{num}]_"
+        index = f"{num}.    "
+
+        if label not in doc or reference in doc:
+            continue
+
+        doc = doc.replace(label, index)
+
+    return doc
+
+
+def _dedent(doc):
+    if not isinstance(doc, str):
+        return doc
+
+    return textwrap.dedent(doc)
+
+
+def _create_op(name):
+    func = _UFuncDispatcher(name)
+    func.__name__ = name
+    doc = getattr(np, name).__doc__
+
+    doc = _remove_unused_reference_labels(_skip_signature(_dedent(doc), name))
+
+    func.__doc__ = (
+        f"xarray specific variant of numpy.{name}. Handles "
+        "xarray objects by dispatching to the appropriate "
+        "function for the underlying array type.\n\n"
+        f"Documentation from numpy:\n\n{doc}"
+    )
+    return func
+
+
+# Auto generate from the public numpy ufuncs
+np_ufuncs = {name for name in dir(np) if isinstance(getattr(np, name), np.ufunc)}
```

Comment:
Can hard code these if preferred?

Reply:
Yes, I would suggest hard coding these if possible, ideally as something like:
```
sin = _unary_ufunc('sin')
add = _binary_ufunc('add')
...
```

The reason to prefer this is that static typing does not evaluate loops or dynamically defined functions. So otherwise `xarray.ufuncs.sin()` will not be recognized as valid by tools like mypy.

Ideally (could be done in a follow-up PR), these functions could be annotated to follow the appropriate type casting rules, so type checkers would know that `xarray.ufuncs.sin(dataset)` returns another Dataset.

In some cases, we use a script to generate all these special methods. I don't think that should be necessary here, but it still may be worth a look to understand the type casting rules:
https://github.com/pydata/xarray/blob/d5f84dd1ef4c023cf2ea0a38866c9d9cd50487e7/xarray/util/generate_ops.py
https://github.com/pydata/xarray/blob/d5f84dd1ef4c023cf2ea0a38866c9d9cd50487e7/xarray/core/_typed_ops.py